### PR TITLE
CI: enhance package detection logic

### PR DESCRIPTION
**Pull request by lipnitsk**
Maintainer: @aparcar

### CODE REVIEW

File: .github/workflows/multi-arch-test-build.yml [hunk 1]
```diff
@@ -35,9 +35,17 @@ jobs:
...
         run: |
           # only detect packages with changed Makefiles
-          PACKAGES="$(git diff --diff-filter=d --name-only origin/master \
-            | grep -E 'Makefile$|test.sh$' | grep -Ev '/files/|/src/' \
-            | awk -F/ '{ print $(NF-1) }' | tr '\n' ' ')"
+          PKG_ROOTS=$(find -name Makefile | sed -e 's@./\(.*\)/Makefile@\1/@')
+          CHANGES=$(git diff --diff-filter=d --name-only origin/master)
+
+          for ROOT in $PKG_ROOTS; do
+            for CHANGE in $CHANGES; do
```
Comment (aparcar) [inline .github/workflows/multi-arch-test-build.yml:42]: Shouldn't we swap these as it's likely always only a single package that changes?

File: .github/workflows/multi-arch-test-build.yml [hunk 2]
```diff
@@ -35,9 +35,17 @@ jobs:
...
+              if [[ "$CHANGE" == "$ROOT"* ]]; then
+                PACKAGES+=$(echo $ROOT | sed -e 's@.*/\(.*\)/@\1 @')
```
Comment (aparcar) [inline .github/workflows/multi-arch-test-build.yml:44]: Doesn't this adds duplicates if two changes happened in a single package.

File: .github/workflows/multi-arch-test-build.yml [hunk 3]
```diff
@@ -35,9 +35,17 @@ jobs:
       - name: Determine changed packages
         run: |
           # only detect packages with changed Makefiles
-          PACKAGES="$(git diff --diff-filter=d --name-only origin/master \
-            | grep -E 'Makefile$|test.sh$' | grep -Ev '/files/|/src/' \
-            | awk -F/ '{ print $(NF-1) }' | tr '\n' ' ')"
+          PKG_ROOTS=$(find -name Makefile | sed -e 's@./\(.*\)/Makefile@\1/@')
```
Comment (aparcar) [inline .github/workflows/multi-arch-test-build.yml:38]: I think we have to make sure that there is no `/src/` in the path as some src folders contain a Makefile.

File: .github/workflows/multi-arch-test-build.yml (hunk 2 above)
Comment (lipnitsk) [reply .github/workflows/multi-arch-test-build.yml:44]: yeah, the `break` below kind of fixes that, by moving onto the next ROOT as soon as a match is found, so duplicates should not match again.

File: .github/workflows/multi-arch-test-build.yml (hunk 3 above)
Comment (lipnitsk) [reply .github/workflows/multi-arch-test-build.yml:38]: Let me test that.

File: .github/workflows/multi-arch-test-build.yml (hunk 1 above)
Comment (lipnitsk) [reply .github/workflows/multi-arch-test-build.yml:42]: Yeah, then the de-dup logic will need some work - I'll take a look

File: .github/workflows/multi-arch-test-build.yml (hunk 2 above)
Comment (aparcar) [reply .github/workflows/multi-arch-test-build.yml:44]: Did you run a benchmark on how long this loop actually takes?

Comment (lipnitsk) [reply .github/workflows/multi-arch-test-build.yml:44]: I think it's negligible - I saw CI kick off and like 10 seconds later it was already past that step. On my machine it takes about 200ms to run the script, but CI is likely slower. Let me add some debug in my temp commit.

Comment (aparcar) [reply .github/workflows/multi-arch-test-build.yml:44]: Okay nevermind then, thanks for checking

File: .github/workflows/multi-arch-test-build.yml (hunk 1 above)
Comment (lipnitsk) [reply .github/workflows/multi-arch-test-build.yml:43]: I don't think it's really worth optimizing this - even in CI this whole thing is done in 1 second or less. Take a look at the CI logs - temp commit prints out `date` before and after this check.

File: .github/workflows/multi-arch-test-build.yml (hunk 3 above)
Comment (lipnitsk) [reply .github/workflows/multi-arch-test-build.yml]: good catch, added a 'grep -v "src/"', seems to fix that.

**Comment by aparcar** [comment]
Excellent thank you very much.

**Pull request merged**